File: apps/windows/TeslaSync.App/feature-views/LiveStatusPill/LiveStatusPill.Model.cs
using System.Globalization;
using TeslaSync.App.Core;
using TeslaSync.App.Core.DataDisplay;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The connection-state the <c>LiveStatusPill</c> reflects — the native analogue of the web
/// <c>StatusLiveState</c> union (<c>'live' | 'reconnecting' | 'offline'</c>) surfaced by
/// <c>useStatusLiveSSE</c> in web/src/features/system/hooks/useStatusLiveSSE.ts. Kept as a dedicated enum
/// (rather than reusing the atomic <c>LiveConnectionState</c>) because this surface's offline tone is the
/// neutral zinc/grey of the web source, whereas the shared live indicator paints its disconnected state
/// danger-red — a different semantic that must not leak in here.
/// </summary>
public enum StatusLiveState
{
    /// <summary>SSE flowing — green dot + "Live" (web <c>'live'</c>).</summary>
    Live,

    /// <summary>Last open errored, re-establishing — pulsing amber + "Reconnecting" (web <c>'reconnecting'</c>).</summary>
    Reconnecting,

    /// <summary>Gave up after backoff — grey + "Offline" (web <c>'offline'</c>).</summary>
    Offline,
}

/// <summary>
/// The render-time data model the <c>LiveStatusPill</c> surface binds to — the native analogue of the web
/// <c>LiveStatusPillProps</c> (<c>{ state, lastUpdateAt, now }</c> in
/// web/src/features/system/components/status/LiveStatusPill.tsx). The web component is purely presentational:
/// the parent /system-status surface owns the <c>useStatusLiveSSE</c> pump and feeds an already-resolved
/// connection <see cref="State"/> plus the <see cref="LastUpdateAt"/> timestamp of the last snapshot, so
/// there is no fetch-driven loading / empty / error / stale branch to reproduce here — the three connection
/// states (including <see cref="StatusLiveState.Offline"/>) ARE the surface's complete state set, exactly as
/// React re-renders the pill with already-resolved props. The web <c>now</c> prop is a render tick rather than
/// data, so it is a projection argument (see <see cref="LiveStatusPillProjection.Project"/>), not a model
/// field. Pure data — no WinUI types — so the projection is unit-tested without a UI host.
/// </summary>
/// <param name="State">The live-pipeline connection state (web <c>state</c>).</param>
/// <param name="LastUpdateAt">When the last snapshot landed, or null before the first (web <c>lastUpdateAt</c>).</param>
public sealed record LiveStatusPillModel(StatusLiveState State, DateTimeOffset? LastUpdateAt)
{
    /// <summary>The initial model — reconnecting with no snapshot yet (the web hook's start state).</summary>
    public static LiveStatusPillModel Connecting { get; } = new(StatusLiveState.Reconnecting, null);

    /// <summary>The offline model — gave up after backoff, no snapshot.</summary>
    public static LiveStatusPillModel Offline { get; } = new(StatusLiveState.Offline, null);
}

/// <summary>
/// The fully projected, render-ready view of a <see cref="LiveStatusPillModel"/> — everything the web
/// component derives before returning JSX: the resolved <see cref="State"/> and its lowercase
/// <see cref="StateToken"/> (web <c>data-status-live-state</c>), the semantic colour <see cref="Tier"/> and
/// its token-backed <see cref="AccentBrushKey"/> (web <c>TONE[state].cls</c> / <c>dot</c>), the Segoe Fluent
/// <see cref="IconGlyph"/> standing in for the web Lucide icon, whether the dot <see cref="Pulse"/>s (web
/// <c>TONE[state].pulse</c>), the localized <see cref="Label"/> (web <c>TONE[state].label</c>), the relative
/// "updated" <see cref="RelativeText"/> (web <c>relative(now, lastUpdateAt)</c>) and the composed Narrator
/// <see cref="AutomationName"/> (web <c>aria-label</c>). Pure data so every value is asserted headlessly.
/// </summary>
/// <param name="State">The connection state, passed through.</param>
/// <param name="StateToken">The lowercase state token (web <c>data-status-live-state</c>): live / reconnecting / offline.</param>
/// <param name="Tier">The semantic colour tier — Success / Warning / Neutral.</param>
/// <param name="AccentBrushKey">The generated design-token brush key the tier resolves to.</param>
/// <param name="IconGlyph">The Segoe Fluent glyph for the state (web Lucide Activity / Wifi / WifiOff).</param>
/// <param name="Pulse">True only while reconnecting — the dot pulses (web <c>pulse</c>).</param>
/// <param name="Label">The localized state label ("Live" / "Reconnecting" / "Offline").</param>
/// <param name="RelativeText">The relative "updated" age, or an em dash before the first update.</param>
/// <param name="AutomationName">The composed Narrator name (web <c>aria-label</c>).</param>
public sealed record LiveStatusPillDisplay(
    StatusLiveState State,
    string StateToken,
    StatusKind Tier,
    string AccentBrushKey,
    string IconGlyph,
    bool Pulse,
    string Label,
    string RelativeText,
    string AutomationName);

/// <summary>
/// Pure projection from a <see cref="LiveStatusPillModel"/> (plus the render clock and the i18n facade) to its
/// <see cref="LiveStatusPillDisplay"/> — the native port of
/// web/src/features/system/components/status/LiveStatusPill.tsx. Reproduces the web derivations exactly: the
/// per-state <c>TONE</c> table (tier colour, pulse flag, icon and label), the <c>relative(now, lastUpdateAt)</c>
/// age tiers (em dash for no update, "just now" under five seconds, then "<c>{s}s ago</c>" / "<c>{m}m ago</c>" /
/// "<c>{h}h ago</c>"), and the composed <c>aria-label</c>. No WinUI types — unit-tested without a UI host.
/// </summary>
public static class LiveStatusPillProjection
{
    /// <summary>Under this many seconds the relative label reads "just now" (web <c>secs &lt; 5</c>).</summary>
    public const int JustNowThresholdSeconds = 5;

    /// <summary>Seconds in a minute — the boundary from "{s}s ago" to "{m}m ago" (web <c>secs &lt; 60</c>).</summary>
    public const int MinuteSeconds = 60;

    /// <summary>Seconds in an hour — the boundary from "{m}m ago" to "{h}h ago" (web <c>secs &lt; 3600</c>).</summary>
    public const int HourSeconds = 3600;

    /// <summary>
    /// The semantic colour tier for <paramref name="state"/>, mirroring the web <c>TONE</c> hues:
    /// live is green (<see cref="StatusKind.Success"/>), reconnecting is amber (<see cref="StatusKind.Warning"/>)
    /// and offline is the neutral zinc/grey (<see cref="StatusKind.Neutral"/>).
    /// </summary>
    public static StatusKind Tier(StatusLiveState state) => state switch
    {
        StatusLiveState.Live => StatusKind.Success,
        StatusLiveState.Reconnecting => StatusKind.Warning,
        _ => StatusKind.Neutral,
    };

    /// <summary>True only while reconnecting — the dot pulses (web <c>TONE[state].pulse</c>).</summary>
    public static bool Pulses(StatusLiveState state) => state == StatusLiveState.Reconnecting;

    /// <summary>The Segoe Fluent glyph for <paramref name="state"/> (web Lucide Activity / Wifi / WifiOff).</summary>
    public static string GlyphFor(StatusLiveState state) => state switch
    {
        StatusLiveState.Live => LiveStatusPillRegistration.ActivityGlyph,
        StatusLiveState.Reconnecting => LiveStatusPillRegistration.WifiGlyph,
        _ => LiveStatusPillRegistration.OfflineGlyph,
    };

    /// <summary>The lowercase state token (web <c>data-status-live-state</c>): live / reconnecting / offline.</summary>
    public static string StateToken(StatusLiveState state) => state switch
    {
        StatusLiveState.Live => "live",
        StatusLiveState.Reconnecting => "reconnecting",
        _ => "offline",
    };

    /// <summary>Resolve the localized state label for <paramref name="state"/> (web <c>TONE[state].label</c>).</summary>
    public static string Label(StatusLiveState state, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        var (key, fallback) = state switch
        {
            StatusLiveState.Live => (LiveStatusPillRegistration.LiveLabelKey, LiveStatusPillRegistration.LiveLabelFallback),
            StatusLiveState.Reconnecting => (LiveStatusPillRegistration.ReconnectingLabelKey, LiveStatusPillRegistration.ReconnectingLabelFallback),
            _ => (LiveStatusPillRegistration.OfflineLabelKey, LiveStatusPillRegistration.OfflineLabelFallback),
        };
        return localizer.GetString(key, fallback);
    }

    /// <summary>
    /// The relative "updated" label for a snapshot that landed at <paramref name="lastUpdateAt"/>, evaluated
    /// against <paramref name="now"/> — the native port of the web <c>relative(now, lastUpdateAt)</c> helper.
    /// A null timestamp is the em dash (web <c>'—'</c>); otherwise the age in whole seconds (floored at zero,
    /// via the shared <see cref="FreshnessLogic.ComputeAge"/> which matches the web
    /// <c>Math.max(0, Math.floor(…/1000))</c>) tiers into "just now" (&lt; 5 s), then "<c>{s}s ago</c>",
    /// "<c>{m}m ago</c>" and "<c>{h}h ago</c>". The "just now" phrase and the "ago" word resolve through the
    /// localizer; the s/m/h unit letters mirror the web source's own literals.
    /// </summary>
    public static string FormatRelative(DateTimeOffset? lastUpdateAt, DateTimeOffset now, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);

        if (FreshnessLogic.ComputeAge(lastUpdateAt, now) is not { } secs)
        {
            return LiveStatusPillRegistration.EmDash;
        }

        if (secs < JustNowThresholdSeconds)
        {
            return localizer.GetString(LiveStatusPillRegistration.JustNowKey, LiveStatusPillRegistration.JustNowFallback);
        }

        var ago = localizer.GetString(LiveStatusPillRegistration.AgoKey, LiveStatusPillRegistration.AgoFallback);
        if (secs < MinuteSeconds)
        {
            return string.Format(CultureInfo.InvariantCulture, "{0}s {1}", secs, ago);
        }

        if (secs < HourSeconds)
        {
            return string.Format(CultureInfo.InvariantCulture, "{0}m {1}", secs / MinuteSeconds, ago);
        }

        return string.Format(CultureInfo.InvariantCulture, "{0}h {1}", secs / HourSeconds, ago);
    }

    /// <summary>Project <paramref name="model"/> into a render-ready display against the render clock.</summary>
    /// <param name="model">The render-time data model (the web props minus the <c>now</c> tick).</param>
    /// <param name="localizer">The i18n facade resolving the label, "just now"/"ago" and the aria-label template.</param>
    /// <param name="now">The render clock — the web <c>now</c> prop driving the relative label.</param>
    public static LiveStatusPillDisplay Project(LiveStatusPillModel model, ILocalizer localizer, DateTimeOffset now)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        var state = model.State;
        var tier = Tier(state);
        var label = Label(state, localizer);
        var relative = FormatRelative(model.LastUpdateAt, now, localizer);
        var ariaTemplate = localizer.GetString(
            LiveStatusPillRegistration.AriaLabelKey,
            LiveStatusPillRegistration.AriaLabelFallback);
        var automationName = string.Format(CultureInfo.CurrentCulture, ariaTemplate, label, relative);

        return new LiveStatusPillDisplay(
            State: state,
            StateToken: StateToken(state),
            Tier: tier,
            AccentBrushKey: StatusResources.AccentBrushKey(tier),
            IconGlyph: GlyphFor(state),
            Pulse: Pulses(state),
            Label: label,
            RelativeText: relative,
            AutomationName: automationName);
    }
}

/// <summary>
/// Canonical metadata for the <c>LiveStatusPill</c> feature surface — the native mirror of the web component at
/// web/src/features/system/components/status/LiveStatusPill.tsx: the stable diagnostics slug, the Segoe Fluent
/// glyphs that stand in for the web Lucide icons, the i18n keys (each with the English fallback the web source
/// renders verbatim) and the literal punctuation the web pill draws. UI-free so the metadata is asserted in
/// tests.
/// </summary>
public static class LiveStatusPillRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11).</summary>
    public const string Slug = "LiveStatusPill";

    /// <summary>Segoe Fluent "activity / pulse" glyph — the web Lucide <c>Activity</c> icon (live).</summary>
    public const string ActivityGlyph = "\uE9D2";

    /// <summary>Segoe Fluent "Wifi" glyph — the web Lucide <c>Wifi</c> icon (reconnecting).</summary>
    public const string WifiGlyph = "\uE701";

    /// <summary>Segoe Fluent "offline" glyph — the native stand-in for the web Lucide <c>WifiOff</c> icon (offline).</summary>
    public const string OfflineGlyph = "\uEB5E";

    /// <summary>i18n key for the live label (web <c>TONE.live.label</c>).</summary>
    public const string LiveLabelKey = "system.liveStatus.label.live";

    /// <summary>English fallback for <see cref="LiveLabelKey"/> — the web literal.</summary>
    public const string LiveLabelFallback = "Live";

    /// <summary>i18n key for the reconnecting label (web <c>TONE.reconnecting.label</c>).</summary>
    public const string ReconnectingLabelKey = "system.liveStatus.label.reconnecting";

    /// <summary>English fallback for <see cref="ReconnectingLabelKey"/> — the web literal (no ellipsis).</summary>
    public const string ReconnectingLabelFallback = "Reconnecting";

    /// <summary>i18n key for the offline label (web <c>TONE.offline.label</c>).</summary>
    public const string OfflineLabelKey = "system.liveStatus.label.offline";

    /// <summary>English fallback for <see cref="OfflineLabelKey"/> — the web literal.</summary>
    public const string OfflineLabelFallback = "Offline";

    /// <summary>i18n key for the "just now" phrase (web <c>'just now'</c>).</summary>
    public const string JustNowKey = "system.liveStatus.justNow";

    /// <summary>English fallback for <see cref="JustNowKey"/> — the web literal (lowercase).</summary>
    public const string JustNowFallback = "just now";

    /// <summary>i18n key for the "ago" word composed into the relative label (web <c>'… ago'</c>).</summary>
    public const string AgoKey = "system.liveStatus.ago";

    /// <summary>English fallback for <see cref="AgoKey"/> — the web literal.</summary>
    public const string AgoFallback = "ago";

    /// <summary>i18n key for the Narrator aria-label template (web <c>aria-label</c>).</summary>
    public const string AriaLabelKey = "system.liveStatus.ariaLabel";

    /// <summary>English fallback for <see cref="AriaLabelKey"/> — the web template (<c>{0}</c>=label, <c>{1}</c>=relative).</summary>
    public const string AriaLabelFallback = "Live status stream: {0}, updated {1}";

    /// <summary>The em dash shown before the first update (web <c>'—'</c>).</summary>
    public const string EmDash = "\u2014";

    /// <summary>The middle dot separating the label from the relative age (web <c>'·'</c>).</summary>
    public const string MiddleDot = "\u00B7";
}

/// <summary>
/// PII-safe diagnostics for the <c>LiveStatusPill</c> surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never the connection state or update time — so
/// a diagnostics line can never leak fleet state. Thread-safe.
/// </summary>
public sealed class LiveStatusPillDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public LiveStatusPillDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=LiveStatusPill</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={LiveStatusPillRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/LiveStatusPill/LiveStatusPillModel.kt
// Pure, framework-free model + projection for the LiveStatusPill feature view — the native analogue of
// everything the web component derives before returning JSX
// (web/src/features/system/components/status/LiveStatusPill.tsx). No Compose, no Android, no HTTP: every
// declaration here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the composable
// a thin render layer.
//
// LiveStatusPill is a purely presentational badge: the web component takes its connection `state`, the
// `lastUpdateAt` stamp, and a `now` tick as props from the /system-status page (which owns the
// `useStatusLiveSSE` pump), so this surface binds NO data hook of its own. As in the sibling BatteryPill /
// AchievementBadge / StatusHeader ports, the cache-then-network lifecycle (loading / error / stale / offline)
// lives on the owning page, not here; modelling those phases would invent behaviour the spec does not have
// (drift). The branches the web source actually defines are the complete state set this surface renders, and
// each is projected here: the three connection states (web `TONE` map: live / reconnecting / offline) and the
// five relative-age buckets the web `relative(now, lastUpdateAt)` produces (em dash / just-now / seconds /
// minutes / hours).
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/LiveStatusPill — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.livestatuspill

import io.teslasync.android.components.datadisplay.FreshnessAge
import io.teslasync.android.components.datadisplay.computeAgeSeconds
import io.teslasync.shared.core.diagnostics.Logger

/**
 * The health of the live SSE stream — the native analogue of the web `StatusLiveState` union
 * (`'live' | 'reconnecting' | 'offline'`) that the `useStatusLiveSSE` pump emits. There is no fourth
 * "unknown" member (unlike the shared [io.teslasync.android.components.datadisplay.LiveConnectionStatus]) so
 * the surface models exactly the three states the web prop can carry.
 */
enum class LiveStatusState {
    /** SSE flowing — green dot + "Live" (web `tone.live`). */
    Live,

    /** Last open errored and a retry is in flight — amber pulsing dot + "Reconnecting" (web `tone.reconnecting`). */
    Reconnecting,

    /** Gave up after backoff — grey dot + "Offline" (web `tone.offline`). */
    Offline,
    ;

    /**
     * The wire token, mirroring the web `data-status-live-state` attribute (and the string the SSE hook
     * emits). Lets the owning state-holder round-trip the state and lets tests pin parity with the web.
     */
    val wire: String
        get() =
            when (this) {
                Live -> "live"
                Reconnecting -> "reconnecting"
                Offline -> "offline"
            }

    companion object {
        /**
         * Maps the live SSE hook's wire token (web `StatusLiveState`) to a state. `"live"` and
         * `"reconnecting"` map verbatim; every other token — including `"offline"` and any unrecognized
         * value — resolves to [Offline], the safe "not live" posture so an unknown stream is never painted
         * as connected.
         */
        fun fromWire(value: String): LiveStatusState =
            when (value) {
                "live" -> Live
                "reconnecting" -> Reconnecting
                else -> Offline
            }
    }
}

/**
 * The fully projected, render-ready view — the native analogue of everything the web component computes
 * before returning JSX. Pure data (no Compose types) so the projection is unit-tested without a UI host;
 * because the composable is a thin render layer, each field here is exactly what it draws.
 *
 * @property state the connection state driving the dot color, icon, label, and pulse (web `TONE[state]`).
 * @property age the relative-age bucket for the "updated X ago" tail (web `relative(now, lastUpdateAt)`).
 * @property pulse whether the status dot animates (web `tone.pulse`) — true only while [LiveStatusState.Reconnecting].
 */
data class LiveStatusPillDisplay(
    val state: LiveStatusState,
    val age: FreshnessAge,
    val pulse: Boolean,
)

/**
 * Pure projection from the web props (`state`, `lastUpdateAt`, `now`) to the render-ready
 * [LiveStatusPillDisplay] — a 1:1 port of the two derivations the web component performs (the `TONE` lookup
 * and the `relative()` helper) before returning JSX.
 */
object LiveStatusPillProjection {
    /** Web `secs < 5`: below this the stream reads "just now". Tighter than the shared 10s FreshnessIndicator window. */
    private const val JUST_NOW_SECONDS: Long = 5

    private const val SECONDS_PER_MINUTE: Long = 60
    private const val SECONDS_PER_HOUR: Long = 3_600

    /** The accessibility sentence join — locale-neutral punctuation, carries no translatable word. */
    private const val SENTENCE_JOIN: String = ". "

    /**
     * The relative-age bucket for the "updated X ago" tail — a 1:1 port of the web
     * `relative(now, lastUpdateAt)`: a null stamp yields [FreshnessAge.Unknown] (the web em dash), `<5s` is
     * just-now, `<60s` is seconds, `<1h` is minutes, and everything else is hours. The web caps at hours — a
     * day-old stream reads "30h ago", never "1d ago" — so this never returns [FreshnessAge.Days] or
     * [FreshnessAge.Weeks]. Reuses the shared [computeAgeSeconds] for the `max(0, floor((now - ts) / 1000))`
     * math (a future stamp clamps to 0 → just-now); the 5-second just-now cutoff is this surface's own.
     */
    fun relativeAge(
        nowMillis: Long,
        lastUpdateAtMillis: Long?,
    ): FreshnessAge {
        val secs = computeAgeSeconds(lastUpdateAtMillis, nowMillis) ?: return FreshnessAge.Unknown
        return when {
            secs < JUST_NOW_SECONDS -> FreshnessAge.JustNow
            secs < SECONDS_PER_MINUTE -> FreshnessAge.Seconds(secs)
            secs < SECONDS_PER_HOUR -> FreshnessAge.Minutes(secs / SECONDS_PER_MINUTE)
            else -> FreshnessAge.Hours(secs / SECONDS_PER_HOUR)
        }
    }

    /** Whether the status dot animates — web `tone.pulse`, set only for the reconnecting state. */
    fun shouldPulse(state: LiveStatusState): Boolean = state == LiveStatusState.Reconnecting

    /** Projects the web props onto the render-ready [LiveStatusPillDisplay]. */
    fun project(
        state: LiveStatusState,
        nowMillis: Long,
        lastUpdateAtMillis: Long?,
    ): LiveStatusPillDisplay =
        LiveStatusPillDisplay(
            state = state,
            age = relativeAge(nowMillis, lastUpdateAtMillis),
            pulse = shouldPulse(state),
        )

    /**
     * The merged accessibility phrase — the native analogue of the web `role="status"` + `aria-label`
     * ("...{label}, updated {rel}"). Composed from the already-localized [stateLabel] and [freshnessPhrase]
     * so it carries no English literal and needs no new catalog key (the surface's allowed-files scope
     * forbids one). Reads e.g. "Live. Last updated: 5s ago" or "Offline. Never updated".
     */
    fun contentDescription(
        stateLabel: String,
        freshnessPhrase: String,
    ): String = stateLabel + SENTENCE_JOIN + freshnessPhrase
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never the
 * connection state or timestamp — so a diagnostics line can never leak a vehicle's live posture.
 */
object LiveStatusPillDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = "LiveStatusPill"

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}
